components/calendar/src/chinese.rs

```diff
@@ -0,0 +1,1213 @@
+
+    /// Get the ordinal lunar month from a code
+    ///
+    /// TODO: The behavior of this fn is calendar specific, but this needs to be implemented in every lunar
+    /// calendar; consider abstracting this function to a Lunar trait
+    fn ordinal_lunar_month_from_code(year: i32, code: MonthCode) -> Option<u8> {
+        if code.0.len() < 3 {
+            return None;
+        }
+        let mid_year = Self::fixed_mid_year_from_year(year);
+        let leap_month = if Self::is_leap_year(year) {
+            Self::get_leap_month_in_year(mid_year)
+        } else {
+            // 14 is a sentinel value, greater than all other months, for the purpose of computation only;
+            // it is impossible to actually have 14 months in a year.
+            14
+        };
+        let bytes = code.0.all_bytes();
+        if bytes[0] != b'M' {
+            return None;
+        }
+        if code.0.len() == 4 && bytes[3] != b'L' {
+            return None;
+        }
+        let mut unadjusted = 0;
+        if bytes[1] == b'0' {
+            if bytes[2] >= b'1' && bytes[2] <= b'9' {
+                unadjusted = bytes[2] - b'0';
+            }
+        } else if bytes[1] == b'1' && bytes[2] >= b'0' && bytes[2] <= b'2' {
+            unadjusted = 10 + bytes[2] - b'0';
+        }
+        if bytes[3] == b'L' {
+            if unadjusted + 1 != leap_month {
+                return None;
+            } else {
+                return Some(unadjusted + 1);
+            }
+        }
+        if unadjusted != 0 {
+            if unadjusted + 1 > leap_month {
+                return Some(unadjusted + 1);
+            } else {
+                return Some(unadjusted);
+            }
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_chinese_new_moon_directionality() {
+        for i in (-1000..1000).step_by(31) {
+            let moment = Moment::new(i as f64);
+            let before = Chinese::chinese_new_moon_before(moment);
+            let after = Chinese::chinese_new_moon_on_or_after(moment);
+            assert!(before < after, "Chinese new moon directionality failed for Moment: {moment:?}, with:\n\tBefore: {before:?}\n\tAfter: {after:?}");
+        }
+    }
+
+    #[test]
+    fn test_chinese_new_year_on_or_before() {
+        let date = Date::try_new_iso_date(2023, 6, 22).expect("Failed to initialize ISO Date");
+        let chinese_new_year = Chinese::chinese_new_year_on_or_before_iso(date);
+        assert_eq!(chinese_new_year.year().number, 2023);
+        assert_eq!(chinese_new_year.month().ordinal, 1);
+        assert_eq!(chinese_new_year.day_of_month().0, 22);
+    }
+
+    #[test]
+    fn test_chinese_from_fixed() {
+        #[derive(Debug)]
+        struct TestCase {
+            fixed: i64,
+            expected_year: i32,
+            expected_month: u32,
+            expected_day: u32,
+        }
+
+        let cases = [
+            TestCase {
+                fixed: -963099,
```

Comment:
Nit: In all calendar systems we should test for pre-epoch behavior (negative years), just enough to make sure it round-trips to RataDie and back. You can add this test later